### ISSUE TITLE
HDF5 and step-62 are not compatible with find_active_cell_around_point()

### ISSUE THREAD

**Issue by dangars** [dealii/dealii#10737]
Step-62 hangs. This related to two issues.
- find_active_cell_around_point() fails in parallel. Even if a point is inside the domain some of the MPI ranks will fail. I think that it is related to #10478 but I'm not sure.
- The `AssertThrow` of find_active_cell_around_point() fails and hangs the program. This is related to #10736

I'm going to prepare a small test to isolate the problem.

**Comment by dangars** [comment]
@masterleinad I have been able to isolate the problem with `AssertThrow` in step-62

It is related to two issues.
- In parallel `find_active_cell_around_point()` fails to find some points in the coarse cells. See test #10747
- `Assert` and `AssertThrow` hang with HDF5. See test #10746

**Comment by dangars** [comment]
The new version of `find_active_cell_around_point()` uses exceptions and that is not compatible with HDF5. See #11050 and #10535

Exceptions combined with parallel HDF5 can freeze deal.II. If one MPI process throws an exception and not the other ones, then deal.II freezes. The reason is that the following happens:
1. On the processor that calls the exception we try to close the parallel MPI file (H5F_try_close) which ultimately results in a barrier (MPI_barrier). This is triggered by the destructor of an HDF5Object.
2. On the processor that does not call the exception we call H5VL_dataset_write() and end up in MPI_Allreduce().

This could happen if an error is found in all the processes except one. For example, the following code could potentially freeze deal.II:
```
HDF5::File data_file(filename,
                          HDF5::File::FileAccessMode::create,
                           mpi_communicator);
{
  auto  dataset_a =
         data_file.create_dataset<double>(dataset_name_a, dataset_dimensions_a);
   dataset_a.write_selection(data_a, coordinates_a);
 }

 AssertThrow(Utilities::MPI::this_mpi_process(mpi_communicator) == 0,
                   ExcInternalError());

{
   auto  dataset_b =
         data_file.create_dataset<double>(dataset_name_b, dataset_dimensions_b);
   dataset_b.write_selection(data_b, coordinates_b);
 }
```

**Comment by masterleinad** [comment]
Isn't this only a problem if the exceptions are not caught and cause the program to abort?

**Comment by dangars** [comment]
I haven't tried to catch the exception of `find_active_cell_around_point()`. I think that using exceptions to communicate that the point is not in an active cell is not very efficient, for this reason I prefer to use an old version of deal.II.

My concert about this issue is that the users that have a code based on step-62 of deal.II 9.2 will experience that their code freezes and does not produce any error even in debug mode.

I have the feeling that there are several people that are using HDF5 and won't know how to fix this. It is really hard to debug this without a parallel debugger.

**Comment by tjhei** [comment]
I don't want to argue against improving find_active_cell_around_point(), but the issue about one MPI rank throwing an exception and it not propagating to the user is a separate issue I feel we should address. I had this in the past in ASPECT, but I worked around this in the past, see https://github.com/dealii/dealii/blob/899b683bb8aaae55e025a55b35a17b24afd035db/source/base/timer.cc#L382-L388 for example.
Basically, we need to avoid MPI communication in destructors that stop the throwing process from exiting.
What is the easiest way to reproduce this hang, @dangars ?

**Comment by dangars** [comment]
@tjhei I agree that the HDF5 interface should avoid MPI communication in destructors that stop the throwing process from exiting.

Thanks for the code. That is a good approach. I will do the same thing.

There are no "explicit class destructors" in the HDF5 interface. The resources are managed  by the variable `hdf5_reference` which is a `std::shared_ptr`. The resource liberation happens inside the lambda/deleter of `std::shared_ptr`. That is where I should place the code with `std::uncaught_exception`.

I'll submit a PR with these modifications and a test in a few days.

**Comment by tjhei** [comment]
Let me know if you need help. If I have a simple test case that hangs, I can make it crash, I think.

**Comment by dangars** [comment]
Thanks! I'll submit the PR with the modifications and I'll let you crash it :)

I think that it should be straightforward to copy your code of timer.cc in the HDF5 interface.

**Comment by dangars** [comment]
@tjhei I started to work on this. I wrote the PR #11339. But it is still not optimal. The test still hangs. I have to do some debugging. Feel free to take a look to the code.

**Comment by peterrum** [comment]
@dangars What is the status here? We a trying to wrap up everything till 18th May (release 9.3). If you don't have time to work on this issue, we can postpone this till after the release.

**Comment by dangars** [comment]
@peterrum I do have time to work on this issue!

The patch is ready see #11310 This patch depends on #11276. Once #11276 is merged it will take me a day to rebase and resubmit the patch.

If #11276 is not going to be ready before May 18th I can try to find another way to fix it.

@luca-heltai Do you think that #11276 will be merged before May 18 or I should take a different approach for `find_active_cell_around_point()` ?

**Comment by luca-heltai** [comment]
I think that https://github.com/dealii/dealii/pull/11276 is ready, once we decide how to proceed with the three failing tests. All comments were addressed.

**Comment by dangars** [comment]
@peterrum The PR that closes this bug is ready for review #11310

**Comment by peterrum** [comment]
@dangars Thank you!

**Comment by dangars** [comment]
:)
